### PR TITLE
Use IPVS for nating instead of iptables

### LINKED ISSUE #852: Proposal: Use IPVS instead of iptables/userland proxy
IPVS is fast and has been part of the kernel since forever.
It would be cool if we could use it instead of iptables (and the userland proxy).
This may also resolve issues seen with hairpin nat on iptables.

See https://github.com/kobolog/gorb as a working daemon that uses ipvs for load balancing...
We'd use it like this except there's only one backend.
This could potentially be utilized when there is some service level concept as well.

### PR DESCRIPTION
**Pull request by cpuguy83**
Using iptables for setting up nat rules has a nasty tendancy to conflict
with user-defined firewall rules, and tend to be invisible to tools like
UFW.
Meanwhile using ipvs for nat is fairly natural and does not require
mucking around in iptables rules.

The one thing we do need to continue using iptables for here is nating
against localhost, however this shouldn't intefere with user-defined
rules unless they are denying localhost (not likely)... and still much
better than having public interfaces here.

Fixes (after re-vendor) docker/docker#4737
Fixes (after re-vendor)  docker/docker#22054
Closes #852

Signed-off-by: Brian Goff cpuguy83@gmail.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
@aboch How come this would fail on `iptables --wait` when trying to inject the `DOCKER` jump rule into `PREROUTING`?

**Comment by aboch** [comment]
@cpuguy83 The test has recently become flacky, we are aware of it. Will invest sometime to fix it after release.
My guess is that it cannot find the DOCKER chain becasue it is somehow executing on a different netns.

File: iptables/iptables.go [hunk 1]
```diff
@@ -157,6 +157,21 @@ func ProgramChain(c *ChainInfo, bridgeName string, hairpinMode, enable bool) err
...
+			"-j", c.Name,
+			"!", "-d", "127.0.0.0/8",
+		}
+		if !Exists(Mangle, "PREROUTING", preroute...) && enable {
+			if err := c.Prerouting(Mangle, Append, preroute...); err != nil {
+				return fmt.Errorf("Failed to inject docker in PREROUTING mangle chain: %s", err)
+			}
+		} else if Exists(Mangle, "PREROUTING", preroute...) && !enable {
+			if err := c.Prerouting(Mangle, Delete, preroute...); err != nil {
+				return fmt.Errorf("Failed to remove docker in PREROUTING nat chain: %s", err)
```
Comment (dongluochen) [inline iptables/iptables.go]: `nat` -> `mangle`

File: iptables/iptables.go [hunk 2]
```diff
@@ -133,12 +133,12 @@ func ProgramChain(c *ChainInfo, bridgeName string, hairpinMode, enable bool) err
 			"--dst-type", "LOCAL",
 			"-j", c.Name}
 		if !Exists(Nat, "PREROUTING", preroute...) && enable {
-			if err := c.Prerouting(Append, preroute...); err != nil {
-				return fmt.Errorf("Failed to inject docker in PREROUTING chain: %s", err)
+			if err := c.Prerouting(Nat, Append, preroute...); err != nil {
+				return fmt.Errorf("Failed to inject docker in PREROUTING nat chain: %s", err)
```
Comment (dongluochen) [inline iptables/iptables.go]: `%s` -> `%v`

**Comment by donnex** [comment]
What is the current status of this issue?

I'm currently using a work around and inject my own iptables chain before the docker rules in order to firewall my traffic. This requires me to run a manual script every time the Docker engine is restarted or the machine is rebooted.

My understanding is that this issue fixes this? Maybe there is another way to deal with it currently?

**Comment by cpuguy83** [comment]
@donnex There's some edge cases not dealt with here. It's also a pretty big change to how Docker handles port forwards which may break existing users.

I'm going to close this, but I think there's another solution to the issue you mentioned.

**Comment by donnex** [comment]
@cpuguy83 Do you have any more information about a viable solution to my problem? I need to firewall traffic to my containers (exposed ports) and limit access by remote IP. I haven't found any solution other than my work-around with injecting firewall rules before the Docker chain.

**Comment by cpuguy83** [comment]
I think the solution will be providing a way to inject rules before the docker chain.

**Comment by aboch** [comment]
> I think the solution will be providing a way to inject rules before the docker chain.

Agree, similar to what #1675 does for the filter table.

**Pull request closed without merging**